### PR TITLE
Change merge set order to topological order

### PR DESCRIPTION
**Pull request by talelbaz**
#1558 
Change merge set order to topological order, update tests accordingly.

### CODE REVIEW

File: domain/consensus/processes/ghostdagmanager/mergeset.go [hunk 1]
```diff
@@ -82,3 +82,42 @@ func (gm *ghostdagManager) sortMergeSet(stagingArea *model.StagingArea, mergeSet
...
+		currentBlueGhostdagData, err := gm.ghostdagDataStore.Get(gm.databaseContext, stagingArea, currentBlue)
+		if err != nil {
+			return nil, err
+		}
+		currentRed := redMergeSet[j]
+		currentRedGhostdagData, err := gm.ghostdagDataStore.Get(gm.databaseContext, stagingArea, currentRed)
+		if err != nil {
+			return nil, err
+		}
+		if gm.Less(currentBlue, currentBlueGhostdagData, currentRed, currentRedGhostdagData) {
```
Comment (someone235) [inline domain/consensus/processes/ghostdagmanager/mergeset.go:111]: Maybe we can compare by blue score instead?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: domain/consensus/processes/coinbasemanager/coinbasemanager.go [hunk 2]
```diff
@@ -141,6 +142,14 @@ func (c *coinbaseManager) coinbaseOutputForRewardFromRedBlocks(stagingArea *mode
 	}, true, nil
 }
 
+func acceptanceDataFromArrayToMap(acceptanceData externalapi.AcceptanceData) map[externalapi.DomainHash]*externalapi.BlockAcceptanceData {
+	acceptanceDataMap := make(map[externalapi.DomainHash]*externalapi.BlockAcceptanceData)
```
Comment (someone235) [inline domain/consensus/processes/coinbasemanager/coinbasemanager.go:146]: You can convert it to `make(map[externalapi.DomainHash]*externalapi.BlockAcceptanceData, len(acceptanceData))` to preallocate the map

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 3]
```diff
@@ -157,8 +156,11 @@ func TestDoubleSpends(t *testing.T) {
 	})
 }
 
-// TestTransactionAcceptance checks that blue blocks transactions are favoured above
-// red blocks transactions, and that the block reward is paid only for blue blocks.
+// TestTransactionAcceptance checks that block transactions are acceptance correctly when the mergeSet is sorted topologically.
```
Comment (someone235) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:159]: acceptance->accepted

Comment (someone235) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:159]: mergeSet -> merge set

File: domain/consensus/processes/ghostdagmanager/mergeset.go [hunk 4]
```diff
@@ -82,3 +82,46 @@ func (gm *ghostdagManager) sortMergeSet(stagingArea *model.StagingArea, mergeSet
 	})
 	return err
 }
+
+func (gm *ghostdagManager) GetSortedMergeSet(stagingArea *model.StagingArea,
```
Comment (someone235) [inline domain/consensus/processes/ghostdagmanager/mergeset.go:86]: Add comment that explains this function return the merge set sorted in a toplogical order

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 5]
```diff
@@ -280,115 +253,76 @@ func TestTransactionAcceptance(t *testing.T) {
...
 		}
-
-		finalTip, err := testConsensus.GetBlock(finalTipHash)
-		if err != nil {
-			t.Fatalf("Error getting finalTip: %+v", err)
-		}
-
-		// We expect the coinbase transaction to pay reward for the selected parent, the
-		// blue block, and bestow the red block reward to the merging block.
-		expectedCoinbase := &externalapi.DomainTransaction{
```
Comment (someone235) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:367]: Why did you remove it?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: domain/consensus/processes/coinbasemanager/coinbasemanager.go (hunk 2 above)
Comment (talelbaz) [reply domain/consensus/processes/coinbasemanager/coinbasemanager.go:146]: Great, thanks.

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 6]
```diff
@@ -156,124 +158,139 @@ func TestDoubleSpends(t *testing.T) {
...
+		blockHashC, _, err := testConsensus.AddBlock([]*externalapi.DomainHash{blockHashB}, nil, nil)
 		if err != nil {
-			t.Fatalf("Error creating fundingBlock3: %+v", err)
+			t.Fatalf("Error creating blockC: %+v", err)
 		}
-		// Add a chain of K blocks above fundingBlock3 so we'll
+		// Add a chain of K blocks above blockHashC so we'll
 		// be able to mine a red block on top of it.
-		tipHash := fundingBlock3Hash
+		chainHash := blockHashC
```
Comment (someone235) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:191]: Maybe call it `chainTipHash`?

File: domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go [hunk 7]
```diff
@@ -156,124 +158,139 @@ func TestDoubleSpends(t *testing.T) {
...
 		if err != nil {
-			t.Fatalf("Error getting UTXOEntry for spendingTransaction1Input0: %s", err)
+			t.Fatalf("Error getting UTXOEntry for transactionFromBlueBlockInput: %s", err)
 		}
-		blueHash, _, err := testConsensus.AddBlock(
-			[]*externalapi.DomainHash{tipHash, redHash},
-			nil,
-			[]*externalapi.DomainTransaction{spendingTransaction1})
+		blueBlockScriptPublicKey := &externalapi.ScriptPublicKey{Script: []byte{3}, Version: 0}
+		blueHash, _, err := testConsensus.AddBlock([]*externalapi.DomainHash{lastBlockInChain, redHash},
```
Comment (someone235) [inline domain/consensus/processes/consensusstatemanager/resolve_block_status_test.go:232]: Maybe you should call it `blueChildOfRedBlock` to differentiate it from all of the other blue blocks?

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — CHANGES_REQUESTED** [review]
You left the `blueBlock` in the DAG diagram in the function comment

**Review by someone235 — APPROVED** [review]
